knodle/trainer/crossweigh_weighing/data_splitting_by_rules.py

```diff
@@ -0,0 +1,289 @@
+
+    return compose_train_n_test_datasets(
+        data_features, signature_id2samples_ids, labels, folds, partitions, other_class_id
+    )
+
+
+def get_rules_sample_ids(rule_matches_z: Union[np.ndarray, sp.csr_matrix]) -> Dict:
+    """
+    This function creates a dictionary {rule id : sample id where this rule matched}. The dictionary is needed as a
+    support tool for faster calculation of train and test sets.
+    :param rule_matches_z: matrix of rule matches as a numpy or sparse matrix (num_samples x num_rules)
+    :return: dictionary {rule_id: [sample_id1, sample_id2, ...]}
+    """
+    if isinstance(rule_matches_z, sp.csr_matrix):
+        rules_samples_ids_dict = {key: set() for key in range(rule_matches_z.shape[1])}
+        for row_id, col_id in zip(*rule_matches_z.nonzero()):       # row_id = sample id, col_id = rule id
+            rules_samples_ids_dict[col_id].add(row_id)
+    else:
+        rules_samples_ids_dict = dict((i, set()) for i in range(0, rule_matches_z.shape[1]))
+        for row_id, row in enumerate(rule_matches_z):
+            rule_ids = np.where(row == 1)[0].tolist()
+            for rule_id in rule_ids:
+                rules_samples_ids_dict[rule_id].add(row_id)
+    return rules_samples_ids_dict
+
+
+def get_signature_sample_ids(rule_matches_z: np.ndarray) -> (Dict, Dict):
+    """
+    This function calculates the signature for each sample (e.g. if rules with ids 1, 5, 7 matched in a sample, a sample
+    signature id is 1_5_7) and id for each signature.
+    :param rule_matches_z: matrix of rule matches (num_samples x num_rules)
+    :return: two dictionaries:
+        - {signature: signature id}
+        - {signature id: [sample_id1, sample_id2, ...]}
+    """
+
+    signature2id, signature_id2samples = {}, {}
+
+    if isinstance(rule_matches_z, sp.csr_matrix):
+        samples_id_rules_dict = {key: [] for key in range(rule_matches_z.shape[0])}
+        for row_id, col_id in zip(*rule_matches_z.nonzero()):
+            samples_id_rules_dict[row_id].append(col_id)
+
+    else:
+        samples_id_rules_dict = dict((i, set()) for i in range(0, rule_matches_z.shape[0]))
+        for row_id, row in enumerate(rule_matches_z):
+            samples_id_rules_dict[row_id] = np.where(row == 1)[0].tolist()
+
+    for sample_id, rules in samples_id_rules_dict.items():
+        signature = "_".join(map(str, sorted(list(rules))))
+
+        if signature in signature2id:
+            signature_id = signature2id.get(signature)
+        else:
+            signature_id = len(signature2id)
+            signature2id[signature] = signature_id
+
+        if signature_id in signature_id2samples:
+            signature_id2samples[signature_id].append(sample_id)
+        else:
+            signature_id2samples[signature_id] = [sample_id]
+
+    return signature2id, signature_id2samples
+
+
+def compose_train_n_test_datasets(
+        data_features: np.ndarray, rule_id2samples_ids: Dict, labels: np.ndarray, folds: int,
```

Comment:
```suggestion
        data_features: np.ndarray, rule_id2samples_ids: Dict, labels: np.ndarray, num_folds: int,
```
to be consistent with other functions?